shared/src/commonMain/kotlin/com/mbta/tid/mbta_app/model/TemporaryTerminalRewriter.kt

```diff
@@ -0,0 +1,254 @@
+        val predictionsAlwaysTypical =
+            routePredictions.isNotEmpty() &&
+                routePredictions.all {
+                    it.routePattern()?.typicality == RoutePattern.Typicality.Typical
+                }
+
+        return isSubway && routeHasAlert && scheduleReplacedTypical && predictionsAlwaysTypical
+    }
+
+    /**
+     * Checks if this route pattern is not typical, points in the same direction as [fullPattern],
+     * contains [stopId], and contains a subsequence of [fullPatternStopIds].
+     */
+    fun RoutePattern.isTruncationOf(
+        fullPattern: RoutePattern,
+        fullPatternStopIds: List<String>,
+        stopId: String
+    ): Boolean {
+        if (typicality == RoutePattern.Typicality.Typical) return false
+        if (directionId != fullPattern.directionId) return false
+        val truncatedPatternStopIds =
+            globalData.trips[representativeTripId]?.stopIds ?: return false
+        return truncatedPatternStopIds.contains(stopId) &&
+            fullPatternStopIds.containsSubsequence(truncatedPatternStopIds)
+    }
+
+    /**
+     * Finds a truncated pattern for [fullPattern] that includes [stopId]. Since at time of writing
+     * we only fetch scheduled trips for the remainder of the day, does not require that the
+     * truncated pattern actually appear in the schedule unless there are multiple plausible
+     * patterns (which could happen if the rating includes both a current disruption and a future
+     * disruption on the same line). If there is no plausible truncated pattern, or if there are
+     * several, does not truncate [fullPattern].
+     */
+    fun truncatedPattern(
+        fullPattern: RoutePattern,
+        fullPatternStopIds: List<String>,
+        stopId: String
+    ): RoutePattern? {
+        // we may be only fetching a subset of the schedule, so we want to consider all the patterns
+        // that could be in the schedule
+        val plausibleTruncatedPatterns =
+            routePatternsByRoute[fullPattern.routeId].orEmpty().filter {
+                it.isTruncationOf(fullPattern, fullPatternStopIds, stopId)
+            }
+
+        return if (plausibleTruncatedPatterns.size < 2) {
+            plausibleTruncatedPatterns.singleOrNull()
+        } else {
+            // if there are multiple truncated patterns with the right direction and stops,
+            // hopefully there's only one that's actually on the schedule
+            plausibleTruncatedPatterns.singleOrNull { truncatedPattern ->
+                schedulesByRoute[fullPattern.routeId].orEmpty().any {
+                    it.routePattern() == truncatedPattern
+                }
+            }
+        }
+    }
+
+    /**
+     * Checks each typical pattern and child stop at [stopPatterns] for truncation, and merges full
+     * patterns into their truncated [NearbyStaticData.StaticPatterns].
+     */
+    fun truncatePatternsAtStop(
```

Comment:
I'm having a hard time wrapping my head around this function - "merges full patterns into their truncated patterns" seems somewhat simpler than what this entails. I think it would help the readability of this and the flow through these functions if it were more functional - instead of mutating truncatedPatternByFullPatternAndStop, could it return the map?

Reply:
I genuinely can't think of a simpler way to merge the full patterns into their truncated patterns than what I'm doing here. Maybe I've just been working on this task for too long; what were you expecting instead?

Reply:
Returning the map makes things a little bit more complicated, because `truncatePatternsAtStop` is called once per stop but `rewritePredictions` is called once per route, but if you think that complexity would make the code more legible I can do that.

Reply:
Maybe I'm confused by what merge is doing here. It seems like we are replacing full patterns with the truncated one, is that right? But I think I'm still not getting the purpose of collapsedPatterns and the relevance of the sorting within it

Reply:
Technically, I guess with the features we have right now, we could be just discarding the full patterns completely, since we're rewriting the predictions' trips to point to the truncated patterns. However, if we did that, when we do something in the future to represent downstream alerts, they won't appear on the temporary terminal headsign, and I think that'd be incorrect. What we actually want to do is merge the full `RoutePattern`s from the full pattern `StaticPatterns` into the truncated `RoutePattern`s' `StaticPatterns`, so that we keep the truncated `StaticPatterns.headsign` but we still have the full `RoutePattern`s. If we don't sort the truncated `StaticPatterns` to the front, then a full `StaticPatterns` might wind up at the front, in which case we'd use that headsign instead, and that's not what we want.

That said, I think if we're merging the full patterns into the truncated `StaticPatterns`, we may not actually need to be also rewriting the predictions, since they'll still be grouped into the truncated headsign. I'm not quite sure if I want to tear that code out, though - rewriting both the `StaticPatterns` and the prediction trip feels like a belt-and-suspenders approach, and since there isn't a temporary terminal right now, we can't easily validate that we only need the belt.

Reply:
If we don't rewrite the trips, then at JFK, Alewife gets discarded as an arrival-only pattern, which is incorrect. (In my tests, the northbound-to-JFK patterns should be being discarded as arrival-only, but the schedules there have slightly inaccurate default data from the `ObjectCollectionBuilder`, which I'm about to fix.)

Reply:
I thought I had finally come up with a cleaner approach for merging things, but it's breaking keeping the ordinary terminal visible and alerting at the disruption boundary, and since the PR is already approved I'm not going to sink even more time into getting that approach working.